### ISSUE TITLE
`--exclusive` nodes do not have their full resources emitted in R

### ISSUE THREAD

**Issue by jameshcorbett** [flux-framework/flux-sched#947]
When match policy is *not* `lonodex` or `hinodex` and nodes are requested exclusively with `--exclusive`, only the minimum matching resource set is included in the generated R, though the node is still allocated exclusively. See e.g. this output on Fluxion, where the system instance uses `lonodex` but single-user instances do not.

```
[corbett8@fluke108:~]$ flux mini alloc -N1 
rsmi_init() failed
[corbett8@fluke6:~]$ flux resource list
     STATE NNODES   NCORES    NGPUS NODELIST
      free      1        4        0 fluke6
 allocated      0        0        0 
      down      0        0        0 
[corbett8@fluke6:~]$ flux mini run -N1 -n1 --exclusive flux start flux resource list
rsmi_init() failed
     STATE NNODES   NCORES    NGPUS NODELIST
      free      1        1        0 fluke6
 allocated      0        0        0 
      down      0        0        0 
[corbett8@fluke6:~]$ flux mini alloc -N1 --exclusive flux resource list
rsmi_init() failed
     STATE NNODES   NCORES    NGPUS NODELIST
      free      1        1        0 fluke6
 allocated      0        0        0 
      down      0        0        0 
```

I think as long as a node is being allocated exclusively, all the resources should be made available to the job.

See also #928.

**Comment by milroy** [comment]
I can reproduce the same behavior in `resource-query`. 

After experimenting with the code, it looks like the issue is caused by the traverser. Unfortunately, the only way I can see so far to emit the "shadow" node resources along with those matched exclusively with the `--exclusive` option is to call  [`add_exclusive_resource_type ("node");`](https://github.com/flux-framework/flux-sched/blob/1f0a19ffe734c641675248bc6191b985b5e818da/resource/policies/dfu_match_policy_factory.cpp#L50) before traversing the graph. Fluxion could be modified to check if the jobspec's node request is `exclusive` before traversing, but that seems like a hack.

Does `--exclusive` apply to other resource types as well?

**Comment by grondo** [comment]
> Does --exclusive apply to other resource types as well?

Technically [yes](https://flux-framework.readthedocs.io/projects/flux-rfc/en/latest/spec_14.html#resources). Can the traverser check to see which resource type is marked as `exclusive` and set `add_exclusive_resource_type (type)` for each exclusive type?

For practical purposes, only the `node` type will be marked exclusive by our current Flux submission tools.

**Comment by milroy** [comment]
> Can the traverser check to see which resource type is marked as exclusive and set add_exclusive_resource_type (type) for each exclusive type?
Yes, I'll have to consider where best to do this. In my test (which fixes the issue) I inserted it before this line: https://github.com/flux-framework/flux-sched/blob/1f0a19ffe734c641675248bc6191b985b5e818da/resource/traversers/dfu_impl.cpp#L1174
but I don't think pulling apart the jobspec within `dfu_impl` is a good design choice.

**Comment by milroy** [comment]
Since the jobspec resources are iterated through in other functions of `dfu_impl` it's probably fine to add checks there. 

Adding the checks and a corresponding `add_exclusive_resource_type (type)` call when the type is `exclusive` causes 41 testsuite tests to fail because of the different emitted resources. I'll need to make sure the fix isn't actually introducing traversal errors, though.

**Comment by milroy** [comment]
I'm closing this issue since the fix in PR #951 is merged.
